include/zephyr/arch/riscv/arch_inlines.h

```diff
@@ -14,7 +14,7 @@
 
 static ALWAYS_INLINE uint32_t arch_proc_id(void)
 {
-	return csr_read(mhartid);
+	return csr_read(mhartid) - CONFIG_RV_BOOT_HART;
```

Comment:
This breaks riscv_machine_timer when RV_BOOT_HART != 0
```c
static uintptr_t get_hart_mtimecmp(void)
{
	return MTIMECMP_REG + (arch_proc_id() * 8);
}
```

Reply:
Hi @ycsin,

I have tested this out on PolarFire SoC and this change breaks things. Moreover, I don't really understand why its needed? 
In a single core application we know that there is a single "Zephyr" CPU and `arch_proc_id()` would return the correct `mhartid `value (as required for the timer driver, etc), irrespective of what hart it boots from.

In an SMP application, we can use `arch_curr_cpu()->id `to return the current "Zephyr" CPU if that is what you are looking for. 

if for some reason you need to know the hartid of the current cpu in an smp application, there is examples of this in the `smp.c`

```c
	for (i = 0; i < CONFIG_MP_MAX_NUM_CPUS; i++) {
		if (_kernel.cpus[i].arch.hartid == hartid) {
			cpu_num = i;
		}
	}
```

Reply:
Yeah I also just came to realize that it shouldn't be done this way, and pulled this back to draft

Reply:
Still trying to verify things internally.

For now, based on my understanding, we seems to have a RISCV SoC that has just a single core and its `mhartid` is non-zero (i.e. 128), which goes against the [RISCV spec](https://riscv.org/wp-content/uploads/2017/05/riscv-privileged-v1.10.pdf):

![Screenshot 2024-03-08 at 7 43 17 PM](https://github.com/zephyrproject-rtos/zephyr/assets/1563974/cfa82006-cad3-404f-8472-da1ae1a6f34a)

`CONFIG_RV_BOOT_HART` doesn't seem to be what we need for this SoC to work as it is meant to skip cores AFAICT, we need something more like an offset instead.

For now I'm considering 2 implementation:
1. Create a new Kconfig (i.e. `CONFIG_RV_HART_OFFSET`) that acts as an offset for `mhartid` that starts from non-zero such that `arch_proc_id()` returns `csr_read(mhartid) - CONFIG_RV_HART_OFFSET`
2. Allow SoC implementation for `arch_proc_id()` - similar to #60771

Reply:
@ycsin @conpax - 

I wonder if we can have a mask applied to the hart id. The normal mask would be -1 (i.e. return all bits of the hart id csr), but then we could have a mask of e.g. 1 for the finicky core we are dealing with.

Something like
```
config RISCV_SOC_HARTID_MASK
    int
    default -1
    help
      Mask to apply to the csr value read in arch_proc_id() for vendor-speficific quirks.

      The default value may be overridden at the SoC level (i.e. this Kconfig is not user-selectable).
```

Reply:
Hey @cfriedt @ycsin 
I would lean more toward the mask idea. However, do you think we could do something with the struct in `include/zephyr/arch/riscv/structs.h`

```c
struct _cpu_arch {
#ifdef CONFIG_USERSPACE
	unsigned long user_exc_sp;
	unsigned long user_exc_tmp0;
	unsigned long user_exc_tmp1;
#endif
#if defined(CONFIG_SMP) || (CONFIG_MP_MAX_NUM_CPUS > 1)
	unsigned long hartid;
	bool online;
#endif
#ifdef CONFIG_FPU_SHARING
	atomic_ptr_val_t fpu_owner;
	uint32_t fpu_state;
#endif
};
```

so the return from arch_proc_id() would be something like: _kernel.cpus[x].arch.hartid or something like that?

Reply:
> @con-pax - and if the hart id condition is false then hard-code `arch_proc_id()` to return 0?

Hey @cfriedt, We still can't hard-code it to return 0 as this will break on platforms that are using a single core other than hart 0. As an example, in an AMP setup where harts 1, 2 and 3 are running Linux and hart 4 is running Zephyr (which is a common setup on PolarFire SoC).

I was thinking more along the lines of "always" having access to hartid (irrespective of SMP). This would lend itself well for if/when Zephyr is to run in S mode. As in, some functionality can be put in place to make an ECALL to get hart id and initialize it for each Zephyr CPU. This way, we don't have to directly read from CSR's each time we need to use arch_proc_id()

The above is just an idea. I also think the mask Idea is good too 😄

Reply:
> so the return from arch_proc_id() would be something like: _kernel.cpus[x].arch.hartid or something like that?

@con-pax isn't this parallel to the mask?